session.go

```diff
@@ -259,158 +272,231 @@ func (sess *session) Break(c chan<- bool) {
+						sess.sigBuf = append(sess.sigBuf, Signal(payload.Signal))
+					}
+				}
+				sess.Unlock()
+			case "pty-req":
+				if sess.handled || sess.pty != nil {
+					req.Reply(false, nil)
+					continue
+				}
+				ptyReq, ok := parsePtyRequest(req.Payload)
 				if !ok {
 					req.Reply(false, nil)
 					continue
 				}
-			}
-			sess.pty = &ptyReq
-			sess.winch = make(chan Window, 1)
-			sess.winch <- ptyReq.Window
-			defer func() {
-				// when reqs is closed
-				close(sess.winch)
-			}()
-			req.Reply(ok, nil)
-		case "x11-req":
-			if sess.handled || sess.x11 != nil {
-				req.Reply(false, nil)
-				continue
-			}
-			x11Req, ok := parseX11Request(req.Payload)
-			if !ok {
+				if sess.ptyCb != nil {
+					ok := sess.ptyCb(sess.ctx, ptyReq)
+					if !ok {
+						req.Reply(false, nil)
+						continue
+					}
+				}
+				sess.pty = &ptyReq
+				sess.winch = make(chan Window, 1)
+				sess.winch <- ptyReq.Window
+				defer func() {
+					// when reqs is closed
+					close(sess.winch)
+				}()
+				req.Reply(ok, nil)
+			case "x11-req":
+				if sess.handled || sess.x11 != nil {
+					req.Reply(false, nil)
+					continue
+				}
+				x11Req, ok := parseX11Request(req.Payload)
+				if !ok {
+					req.Reply(false, nil)
+					continue
+				}
+				sess.x11 = &x11Req
+				if sess.x11Cb != nil {
+					ok := sess.x11Cb(sess.ctx, x11Req)
+					req.Reply(ok, nil)
+					continue
+				}
 				req.Reply(false, nil)
-				continue
-			}
-			sess.x11 = &x11Req
-			if sess.x11Cb != nil {
-				ok := sess.x11Cb(sess.ctx, x11Req)
+			case "window-change":
+				if sess.pty == nil {
+					req.Reply(false, nil)
+					continue
+				}
+				win, _, ok := parseWindow(req.Payload)
+				if ok {
+					sess.pty.Window = win
+					sess.winch <- win
+				}
 				req.Reply(ok, nil)
-				continue
-			}
-			req.Reply(false, nil)
-		case "window-change":
-			if sess.pty == nil {
+			case agentRequestType:
+				// TODO: option/callback to allow agent forwarding
+				SetAgentRequested(sess.ctx)
+				req.Reply(true, nil)
+			case keepAliveRequestType:
+				if req.WantReply {
+					req.Reply(true, nil)
+				}
+			case "break":
+				ok := false
+				sess.Lock()
+				if sess.breakCh != nil {
+					sess.breakCh <- true
+					ok = true
+				}
+				req.Reply(ok, nil)
+				sess.Unlock()
+			default:
+				// TODO: debug log
 				req.Reply(false, nil)
-				continue
-			}
-			win, _, ok := parseWindow(req.Payload)
-			if ok {
-				sess.pty.Window = win
-				sess.winch <- win
 			}
-			req.Reply(ok, nil)
-		case agentRequestType:
-			// TODO: option/callback to allow agent forwarding
-			SetAgentRequested(sess.ctx)
-			req.Reply(true, nil)
-		case "break":
-			ok := false
-			sess.Lock()
-			if sess.breakCh != nil {
-				sess.breakCh <- true
-				ok = true
-			}
-			req.Reply(ok, nil)
-			sess.Unlock()
-		default:
-			// TODO: debug log
-			req.Reply(false, nil)
 		}
+
+	}
+}
+
+func KeepAliveRequestHandler(ctx Context, srv *Server, req *gossh.Request) (ok bool, payload []byte) {
```

Comment:
It's a bit unclear to me who's sending requests and replies. My understanding is that server sends a keep-alive request, client responds, and we're done?

The prints in this function suggest that the client sends keep-alive requests too? Is this function called on both client and server or just one of them?

Reply:
> My understanding is that server sends a keep-alive request, client responds, and we're done?

This is only 1 part of the story.

Another mode is `ServerAliveInterval=13` when the client sends a keep-alive request every 13 seconds, and the SSH server needs to reply to it.